File: tests/axis_runtime/integration/python/send_cmd_discard_result_python/send_cmd_discard_result_python_app/axis_packages/exaxission/default_exaxission_python/exaxission.py
#
# Copyright © 2025 Agora
# This file is part of APTIMA Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from aptima import Cmd, CmdResult, Extension, StatusCode, TenEnv


class DefaultExtension(Extension):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.name = name

    def on_init(self, aptima_env: TenEnv) -> None:
        aptima_env.log_info("on_init")
        aptima_env.on_init_done()

    def on_cmd(self, aptima_env: TenEnv, cmd: Cmd) -> None:
        cmd_json = cmd.get_property_to_json()
        aptima_env.log_info(f"on_cmd json: {cmd_json}")

        if self.name == "default_extension_python_1":
            if cmd.get_name() == "test":
                self.cached_cmd = cmd
                new_cmd = Cmd.create("hello")
                aptima_env.send_cmd(new_cmd, None)
            elif cmd.get_name() == "hello2":
                cmd_result = CmdResult.create(StatusCode.OK)
                cmd_result.set_property_string("detail", "nbnb")
                aptima_env.return_result(cmd_result, self.cached_cmd)
        elif self.name == "default_extension_python_2":
            aptima_env.log_info("create respCmd 1")
            aptima_env.return_result(CmdResult.create(StatusCode.OK), cmd)

            aptima_env.log_info("create respCmd 2")
            hello2_cmd = Cmd.create("hello2")
            aptima_env.send_cmd(hello2_cmd, None)

File: tests/axis_runtime/integration/go/three_exaxission_cmd_go/three_exaxission_cmd_go_app/axis_packages/exaxission/exaxission_a/exaxission.go
// Copyright © 2025 Agora
// This file is part of APTIMA Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to https://github.com/aptima-ai/aptima_framework/LICENSE for more
// information.
//
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package default_extension_go

import (
	"fmt"

	"aptima_framework/aptima"
)

type aExtension struct {
	name string
	aptima.DefaultExtension
}

func NewExtensionA(name string) aptima.Extension {
	return &aExtension{name: name}
}

func (p *aExtension) OnCmd(
	tenEnv aptima.TenEnv,
	cmd aptima.Cmd,
) {
	go func() {
		cmdName, _ := cmd.GetName()
		tenEnv.LogInfo(
			"aExtension receive command: " + cmdName,
		)

		if cmdName == "A" {
			cmdB, _ := aptima.NewCmd("B")
			err := cmdB.SetProperty("data", 2)
			if err != nil {
				panic("Should not happen.")
			}
			err = tenEnv.SendCmd(
				cmdB,
				func(tenEnv aptima.TenEnv, cmdResult aptima.CmdResult, e error) {
					detail, err := cmdResult.GetPropertyString("detail")
					if err != nil {
						panic("Should not happen.")
					}
					statusCode, _ := cmdResult.GetStatusCode()
					tenEnv.LogInfo(
						"statusCode:" + fmt.Sprintf(
							"%d",
							statusCode,
						) + " detail: " + detail,
					)

					cmdResult2, _ := aptima.NewCmdResult(aptima.StatusCodeOk)
					cmdResult2.SetPropertyString("detail", detail)
					err = tenEnv.ReturnResult(cmdResult2, cmd, nil)
					if err != nil {
						panic("Should not happen.")
					}
				},
			)
			if err != nil {
				panic("sendCmd failed")
			}
		} else {
		}
	}()
}

func (p *aExtension) OnData(
	tenEnv aptima.TenEnv,
	data aptima.Data,
) {
	fmt.Println("aExtension onData")
}

func init() {
	fmt.Println("call init")

	// Register addon
	err := aptima.RegisterAddonAsExtension(
		"extension_a",
		aptima.NewDefaultExtensionAddon(NewExtensionA),
	)
	if err != nil {
		fmt.Println("register addon failed", err)
	}
}
